File: strings/205. Isomorphic Strings/isIsomorphic.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function(s, t) {
    let char_map = {};

    for (let i = 0; i < s.length; i++) {
        // If key doesn't exist in the obj: !Object.keys(char_map).includes(s[i])
        if (!(s[i] in char_map)) {
            if (Object.values(char_map).includes(t[i])) {
                return false;
            }
            char_map[s[i]] = t[i];
        } else {
            if (t[i] !== char_map[s[i]]) {
                return false;
            }
        }
    }

    return true;
};

File: strings/205. Isomorphic Strings/isIsomorphic.py
class Solution(object):
    def isIsomorphic(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        # Hash map holding characters from s as keys mapped to character from t as values (s: t)
        char_map = {}

        for i in range(len(s)):
            if s[i] not in char_map:  # Check if s[i] doesn't exist as a key
                if t[i] in char_map.values():  # t[i] shouldn't exist since they should be mapped together
                    return False  # If it does, it means t[i] is mapped to a different char from s
                    """ s = "badc"
                        t = "baba"
                        -> i = 2: s[i] = d, doesn't exist as key, but t[i] = b exists as value and is mapped to b,
                            return False
                    """
                char_map[s[i]] = t[i]  # Both don't exist, create a new pair
            else:
                # Since s[i] exists as key it should be mapped to t[i] as value
                if t[i] != char_map[s[i]]:
                    return False
        
        return True